Enable kernal to accurately detect amount of memory installed

JimmyDansbo opened the pull request:
ROM was relying on the fact that the emulator wraps around to bank 0 when no more banks of memory are present. Real hardware does not have this functionality instead, writes to empty banks are not stored anywhere. For this reason I have added a bit of code that checks that a write to a bank can actually be read back, then it compares it with value on bank 0 to ensure continued support in the emulator as well.
I have tested the ROM in latest emulator and on unoffical hardware designed by Joe Burks (Wavicle)

JimmyDansbo committed:
Enable kernal to accurately detect amount of memory installed
ROM was relying on the fact that the emulator wraps around to
bank 0 when no more banks of memory are present. Real hardware
does not have this functionality instead, writes to empty banks
are not stored anywhere. For this reason I have added a bit of
code that checks that a write to a bank can actually be read
back, then it compares it with value on bank 0 to ensure
continued support in the emulator as well.
I have tested the ROM in latest emulator and on unoffical
hardware designed by Joe Burks (Wavicle)

JimmyDansbo committed:
Added comment explaining ASL to INC switch.
Switched to INC instead of ASL to ensure that odd memory
sizes such as 1536K are also detected. With ASL it would
only be possible to detect 1, 2, 4, 8, 16, 32, 64, 128 or
256 banks which means 1536K could not be detected.

mooinglemur commented:
I feel that this check could rely too much on reading a specific value out of RAM without checking that the RAM actually changed from its previous value.  Open bus behavior could result in a false positive for the availability of a bank, on average 1 out of 256 times, of course depending on the data bus's previous values.

It seems like we should do multiple write-reads of each bank's $a000 for more robustness

I think the original check had similar deficiencies, though, so this is nothing new.  But rather than showing, for instance, 1024K of RAM on a 512K system, it would show 520K.

mooinglemur commented:
Perhaps something like this would be more robust?  Just an idea :)
```
lda $a000
eor #$ff
sta $a000
cmp $a000
bne whatever
; put it back the way it was
eor #$ff
sta $a000
```

JimmyDansbo commented:
I think it may be a question of reliability vs. size, but I agree that relying on the, more or less, random value obtained by reading $a000 on bank0 and incrementing it is not optimal.

ZeroByteOrg commented:
Would it make sense to have some "sentinel value" in $00:A000 and then subsequent writes to other banks should be different than both the initial value AND the sentinel value?
I.e. read $00:A000 and then $01:A000. INC $01:A000, read $00:A000 (to "cleanse the palette of the bus) and then re-read $01:A000. If new value is returned, DEC $01:A000 and then move to next bank.
I'd imagine that banks could be checked as $00, $40, $80, $C0 and be safe to determine RAM amount, as the system uses 512K chips in 4 sockets.

JimmyDansbo commented:
I have been trying to keep the code as close to the original as possible without using more ROM space than absolutely necessary. I believe that my current entry adds 3 bytes to the ROM. I think the idea of "cleanse the pallette of the bus" is good, but I can't figure a way of doing it without adding quite a bit of code.
Also we need to keep in mind that the emulator will just wrap around to bank zero if there is no RAM at the tested bank, i.e. Startet without the -ram option, banks $40, $80 & $C0 wrap to bank $00. With -ram 1024, banks $80 & $C0 wrap to $00.
As the system is not supported with less than 512KB of memory, it would actually suffice to test banks $40, $80 and $C0 unless we want the kernal to be able to accurately detect the amount of memory available even if it is an "unsupport" amount.

indigodarkwolf commented:
If the emulator's behavior is different from the hardware here, we should change the emulator's behavior. This should be a simple change.

JimmyDansbo commented:
Wavicle tested on prototype3 and at the moment the kernal believes that it has 2048KB even though there is only 512KB. I believe that emulator should be changed to not expect any wrap-around at all and I will try to write some new detection code that uses the EOR function to test banks $00, $40, $80 & $C0

JimmyDansbo committed:
Complete rewrite of ram bank detection code.
The code has been rewritten to do read, xor, store, compare
on two address in a bank as well as compare with bank 0 to
check for wrap-around.

JimmyDansbo committed:
Another rewrite as suggested by stefan-b-jakobsson
Original bank-detector is kept, almost intact, but another
loop is added to ensure that the correct amount of memory
is detected when there is no wrap-around to bank0

JimmyDansbo committed:
RAM bank does not need to bet set to 0
Realized that a few lines later, RAM bank is set to 1

irmen commented:
@JimmyDansbo it may be easier for you if you're not making this PR on your master branch but instead use a feature branch to stick the changes on... ?   (now you can't mix it with other things that are happening on the master branch perhaps.)

JimmyDansbo commented:
@irmen I am not really familiar with how to do things on github, but I do see your point. In the future I will see if I can create a branch before creating a pull request. For the moment I don't really have anything else I can contribute to the ROM so lets just hope that this gets merged before I find something else I would like to work on.

JimmyDansbo commented:
Closing Pull request as it has been merged
